language-plutus-core/src/Language/PlutusCore/Analysis/Definitions.hs

```diff
@@ -87,37 +86,38 @@ addUsage
     -> ScopeType -- ^ The scope type
     -> m ()
 addUsage n newUse tpe = do
-    let u = coerce $ n ^. unique
     let use = ScopedLoc tpe newUse
 
-    d@(def, uses) <- lookupDef u
-    checkCoherency u use d
-    checkDefined u use d
-
-    modify $ IM.insert (unUnique u) (def, Set.insert use uses)
+    d@(def, uses) <- lookupDef n
+    checkCoherency n use d
+    checkDefined n use d
+    modify $ insertByNameIndex n (def, Set.insert use uses)
 
 checkDefined
-    :: (MonadWriter [UniqueError a] m)
-    => Unique -- ^ The unique
+    :: (HasUnique n u,
+        MonadWriter [UniqueError a] m)
+    => n -- ^ The variable
```

Comment:
I changed things here a bit: various checks now receive names rather than uniques. It's not strictly necessary to change this file, but I think it would be good anyway to have something like

```haskell
data ScopedName tyname name a
    = ScopedTyName (tyname a)
    | ScopedName (name a)
```

and turn ` UniqueError` into `NameError`. Annotations and actual textual representations of names is a useful debug information.

Reply:
Yes, I agree. I wasn't totally happy with the design of this file.